spark3.0/src/main/scala/com/splicemachine/spark/jms/JmsDataSourceRelation.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 https://github.com/jksinghpro/spark-jms
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.splicemachine.spark.jms
+
+import java.util.Collections
+import javax.jms._
+
+import org.apache.activemq.ActiveMQConnectionFactory
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.ScalaReflection
+import org.apache.spark.sql.{Row, SQLContext}
+import org.apache.spark.sql.sources.{BaseRelation, TableScan}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+
+import scala.beans.BeanProperty
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Created by exa00015 on 17/12/18.
+  */
+class JmsDatasourceRelation(override val sqlContext: SQLContext, parameters: Map[String, String]) extends BaseRelation with TableScan with Serializable {
+
+  lazy val RECIEVER_TIMEOUT = parameters.getOrElse("reciever.timeout","3000").toLong
+
+
+  override def schema: StructType = {
+    ScalaReflection.schemaFor[JmsMessage].dataType.asInstanceOf[StructType]
```

Comment:
What a trick! I think `Encoders.product[JmsMessage].schema` could work. If so, use it below to create a DataFrame out of `JmsMessage`s.